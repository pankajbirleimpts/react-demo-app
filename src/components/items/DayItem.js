import React, { Component } from "react";
import { withRouter, Link } from "react-router-dom";
import { connect } from "react-redux";
import { Formik, Form, Field, ErrorMessage } from "formik";
import * as Yup from "yup";
import { Grid } from "semantic-ui-react";
import SemanticDatepicker from "react-semantic-ui-datepickers";
import "react-semantic-ui-datepickers/dist/react-semantic-ui-datepickers.css";

import {
  getAllItems,
  addDayItem,
  updateDayItem,
  getDayItem
} from "../../actions/ItemAction";
import { langs } from "../../config";
import { Loader } from "../common";
import "./dayitem.css";

class DayItem extends Component {
  state = {
    id: "",
    initialValues: {
      itemId: "",
      itemDetails: null,
      date: "",
      totalQuantity: 0,
      remainingQuantity: 0
    }
  };

  componentDidMount() {
    this.props.getAllItems();
    const { id } = this.props.match.params;
    if (id) {
      this.props.getDayItem(id, response => {
        console.log("response", response);
        this.setState({
          initialValues: response,
          id
        });
      });
    }
  }
  /**
   * @formSubmitHandler
   * @desc: Submit the signup form
   */
  formSubmitHandler = values => {
    console.log("formSubmitHandler values ", values);
    values.itemDetails = this.props.item.allItems.find(
      val => val.id === values.itemId
    );
    if (this.state.id !== "") {
      // Update item
      this.props.updateDayItem(
        this.state.id,
        values,
        () => {
          this.props.history.replace("/day-items");
        },
        true
      );
    } else {
      // Add Item
      this.props.addDayItem(values, () => {
        this.props.history.replace("/day-items");
      });
    }
  };

  /**
   * @formValidation
   * @desc: form validation
   */
  formValidation = () => {
    return Yup.object().shape({
      date: Yup.string().required(langs.messages.REQUIRED),
      itemId: Yup.string().required(langs.messages.REQUIRED),
      totalQuantity: Yup.number().required(langs.messages.REQUIRED)
    });
  };

  /**
   * @renderForm
   * @desc: rednder form
   */
  renderForm = ({ values, setFieldValue }) => {
    return (
      <Form noValidate className="ui form">
        <div className="field">
          <label>Date *</label>
          <SemanticDatepicker
            format="DD-MMM-YYYY"
            name="date"
            onChange={(e, { name, value }) => {
              setFieldValue("date", value);
            }}
          />
          <ErrorMessage component="p" name="date" className="red" />
        </div>
        <div className="field">
          <label>Item *</label>
          <Field name="itemId" as="select">
            <option value="">Please Select</option>
            {this.props.item.allItems.map(val => {
              return (
                <option key={val.id} value={val.id}>
                  {val.itemName}
                </option>
              );
            })}
          </Field>
          <ErrorMessage component="p" name="itemId" className="red" />
        </div>
        <div className="field">
          <label>Quantity *</label>
          <Field name="totalQuantity" type="number" />
          <ErrorMessage component="p" name="totalQuantity" className="red" />
        </div>
        <Link to="/day-items" className="ui button">
          Back to Items List
        </Link>
        <button className="ui button primary" type="submit">
          {this.state.id !== "" ? "Update" : "Submit"}
        </button>
      </Form>
    );
  };

  render() {
    return (
      <Grid>
        <Loader isLoading={this.props.item.isLoading} />
        <Grid.Row centered className="add-day-item-container">
          <Grid.Column width="8">
            <h2> {this.state.id !== "" ? "Update" : "Add"} Day Item</h2>
            <Formik
              enableReinitialize
              initialValues={this.state.initialValues}
              validationSchema={this.formValidation()}
              onSubmit={values => {
                this.formSubmitHandler(values);
              }}
            >
              {this.renderForm}
            </Formik>
          </Grid.Column>
        </Grid.Row>
      </Grid>
    );
  }
}

function mapStateToProp({ item }) {
  return {
    item
  };
}

export default connect(mapStateToProp, {
  addDayItem,
  updateDayItem,
  getDayItem,
  getAllItems
})(withRouter(DayItem));
