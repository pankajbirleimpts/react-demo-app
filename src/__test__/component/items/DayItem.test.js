import React from 'react';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { BrowserRouter } from 'react-router-dom';
import DayItem, { UnconnectedDayItem } from '../../../components/items/DayItem';
import { storeFactory, findByAttr } from '../../test/TestUtils';
import SemanticDatepicker from "react-semantic-ui-datepickers";
jest.mock('react-semantic-ui-datepickers');



const setup = (props) => {
    const wrapper = shallow(
        <UnconnectedDayItem {...props} />,
    )
    return wrapper;
};

describe('Add item Flow', () => {
    const props = {
        item: {
            isLoading: false,
            allItems: [{
                id: 1,
                itemName: "Pahe"
            }]
        },
        history: {
            replace: jest.fn(),
        },
        getAllItems: jest.fn(),
        getDayItem: jest.fn(),
        updateDayItem: jest.fn(),
        addDayItem: jest.fn(),
        match: {
            params: {
                id: undefined,
            }
        }
    }
    test('Snapshot of DayItem component', () => {
        const wrapper = setup(props);
        expect(wrapper).toMatchSnapshot();
    });


    test('Check the isLoading in props', () => {
        const wrapper = setup(props);
        const componentProps = wrapper.instance().props;
        expect(componentProps.item.isLoading).toBe(props.item.isLoading);
    });

    test('Check getDayItem action should not be called props', () => {
        const wrapper = setup(props);
        wrapper.instance().componentDidMount();
        expect(props.getDayItem).not.toHaveBeenCalled();
    });

    test('Check page titile should be `Add item`', () => {
        const wrapper = setup(props);
        const pageTitle = findByAttr(wrapper, 'page-heading');
        expect(pageTitle.text()).toEqual('Add Day Item');
    });

    test('Checking the form values ', async () => {
        const formValues = {
            date: "17-Feb-2020",
            itemId: 1,
            totalQuantity: 5,
        };

        const wrapper = mount(<BrowserRouter><UnconnectedDayItem {...props} /></BrowserRouter>);

        const dateInput = findByAttr(wrapper, 'date-input').find('input').first();
        const itemIdInput = findByAttr(wrapper, 'itemId-input').find('select').first();
        const totalQuantityInput = findByAttr(wrapper, 'totalQuantity-input').find('input').first();

        await act(async () => {
            dateInput.simulate('change', { persist: () => { }, target: { name: 'date', value: formValues.date } });
        });

        await act(async () => {
            itemIdInput.simulate('change', { persist: () => { }, target: { name: 'itemId', value: formValues.itemId } });
        });

        await act(async () => {
            totalQuantityInput.simulate('change', { persist: () => { }, target: { name: 'totalQuantity', value: formValues.totalQuantity } });
        });

        await act(async () => {
            wrapper.find('form').simulate('submit');
        });
        expect(props.addDayItem).toHaveBeenCalled();

    });
});

 describe('Update item Flow', () => {
    const props = {
        item: {
            isLoading: false,
            allItems: [{
                id: 1,
                itemName: "Pahe"
            }]
        },
        history: {
            replace: jest.fn(),
        },
        getAllItems: jest.fn(),
        getDayItem: jest.fn(),
        updateDayItem: jest.fn(),
        addDayItem: jest.fn(),
        match: {
            params: {
                id: 101,
            }
        }
    }

    test('Check the isLoading in props', () => {
        const wrapper = setup(props);
        const componentProps = wrapper.instance().props;
        expect(componentProps.item.isLoading).toBe(props.item.isLoading);
    });

    test('Check getDayItem action should be called props', () => {
        const wrapper = setup(props);
        wrapper.instance().componentDidMount();
        expect(props.getDayItem).toHaveBeenCalled();
    });

    test('Check page titile should be `Update item`', () => {
        const wrapper = setup(props);
        wrapper.instance().setState({ id: props.match.params.id });
        const pageTitle = findByAttr(wrapper, 'page-heading');
        expect(pageTitle.text()).toEqual('Update Day Item');
    });

    test('Checking the form values ', async () => {

        const formValues = {
            date: "17-Feb-2020",
            itemId: 1,
            totalQuantity: 5,
        };

        const wrapper = mount(<BrowserRouter><UnconnectedDayItem {...props} /></BrowserRouter>);

        wrapper.instance().setState({
            id: props.match.params.id,
            initialValues: formValues,
        });
        const dateInput = findByAttr(wrapper, 'date-input').find('input').first();
        const itemIdInput = findByAttr(wrapper, 'itemId-input').find('select').first();
        const totalQuantityInput = findByAttr(wrapper, 'totalQuantity-input').find('input').first();

        await act(async () => {
            dateInput.simulate('change', { persist: () => { }, target: { name: 'date', value: formValues.date } });
        });

        await act(async () => {
            itemIdInput.simulate('change', { persist: () => { }, target: { name: 'itemId', value: formValues.itemId } });
        });

        await act(async () => {
            totalQuantityInput.simulate('change', { persist: () => { }, target: { name: 'totalQuantity', value: formValues.totalQuantity } });
        });

        await act(async () => {
            wrapper.find('form').simulate('submit');
        });
        expect(props.updateDayItem).toHaveBeenCalled();

  });
});