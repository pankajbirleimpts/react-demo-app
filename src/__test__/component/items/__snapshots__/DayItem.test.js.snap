// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add item Flow Snapshot of DayItem component 1`] = `
<Grid>
  <Loading
    isLoading={false}
  />
  <GridRow
    centered={true}
    className="add-day-item-container"
  >
    <GridColumn
      width="8"
    >
      <h2
        data-test="page-heading"
      >
        Add
         Day Item
      </h2>
      <Formik
        enableReinitialize={true}
        initialValues={
          Object {
            "date": "",
            "itemDetails": null,
            "itemId": "",
            "remainingQuantity": 0,
            "totalQuantity": 0,
          }
        }
        onSubmit={[Function]}
        validationSchema={
          ObjectSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_defaultDefault": [Function],
            "_deps": Array [],
            "_excludedEdges": Array [],
            "_exclusive": Object {},
            "_mutate": undefined,
            "_nodes": Array [
              "totalQuantity",
              "itemId",
              "date",
            ],
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "object",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "fields": Object {
              "date": StringSchema {
                "_blacklist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_conditions": Array [],
                "_deps": Array [],
                "_exclusive": Object {
                  "required": true,
                },
                "_mutate": undefined,
                "_options": Object {
                  "abortEarly": true,
                  "recursive": true,
                },
                "_type": "string",
                "_typeError": [Function],
                "_whitelist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "tests": Array [
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
              },
              "itemId": StringSchema {
                "_blacklist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_conditions": Array [],
                "_deps": Array [],
                "_exclusive": Object {
                  "required": true,
                },
                "_mutate": undefined,
                "_options": Object {
                  "abortEarly": true,
                  "recursive": true,
                },
                "_type": "string",
                "_typeError": [Function],
                "_whitelist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "tests": Array [
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
              },
              "totalQuantity": NumberSchema {
                "_blacklist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_conditions": Array [],
                "_deps": Array [],
                "_exclusive": Object {
                  "required": true,
                },
                "_mutate": undefined,
                "_options": Object {
                  "abortEarly": true,
                  "recursive": true,
                },
                "_type": "number",
                "_typeError": [Function],
                "_whitelist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "tests": Array [
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
              },
            },
            "tests": Array [],
            "transforms": Array [
              [Function],
            ],
          }
        }
      >
        <Component />
      </Formik>
    </GridColumn>
  </GridRow>
</Grid>
`;
